### PR TITLE
improve XInput support

### PR DESCRIPTION
**Pull request by benoit-pierre**
Add support for:
* `grab_device/ungrab_device` requests
* `grab_keycode/ungrab_keycode` requests
* `query_device` request
*  `ButtonPress/ButtonRelease/KeyPress/KeyRelease/Motion` device events
* `DeviceChanged` event

### CODE REVIEW

File: Xlib/ext/xinput.py [hunk 1]
```diff
@@ -138,10 +147,35 @@
...
+    structvalues = 2
+
+    def check_value(self, value):
+        return value
+
+    def parse_value(self, value, display):
+        integral, frac = value
+        ret = float(integral)
+        # optimised math.ldexp(float(frac), -32)
+        ret += float(frac) * (1.0 / (1 << 32))
```
Comment (vasily-v-ryabov) [inline Xlib/ext/xinput.py]: Shouldn't it be `float(1 << 32)` like for `FP1616`?

File: Xlib/ext/xinput.py [hunk 2]
```diff
@@ -170,12 +204,41 @@ def query_version(self):
...
+
+        if isinstance(val, integer_types):
+            # We need to build a "binary mask" that (as far as I can tell) is
+            # encoded in native byte order from end to end.  The simple case is
+            # with a single unsigned 32-bit value, for which we construct an
+            # array with just one item.  For values too big to fit inside 4
+            # bytes we build a longer array, being careful to maintain native
+            # byte order across the entire set of values.
+            if sys.byteorder == 'little':
+                fn = lambda v: mask_seq.insert(0, v)
```
Comment (vasily-v-ryabov) [inline Xlib/ext/xinput.py]: This is a QC minor warning: [Avoid assigning a lambda expression to a variable](https://www.quantifiedcode.com/app/issue_class/6803rnFX)

File: Xlib/ext/xinput.py (hunk 1 above)
Comment (benoit-pierre) [inline Xlib/ext/xinput.py]: No (Note the code is copied from Xorg).

File: Xlib/ext/xinput.py (hunk 2 above)
Comment (benoit-pierre) [inline Xlib/ext/xinput.py]: Amended.

File: Xlib/ext/xinput.py (hunk 1 above)
Comment (vasily-v-ryabov) [inline Xlib/ext/xinput.py]: OK, it's just for peace of mind because we have no tests for extensions.

**Comment by benoit-pierre** [comment]
I'm going to look at removing those string interpolations QuantifiedCode is warning about, please don't merge just yet.

**Comment by benoit-pierre** [comment]
Done.

**Comment by vasily-v-ryabov** [comment]
Cool! Thanks.

**Pull request merged**